src/osx_clipboard.rs

```diff
@@ -12,74 +12,71 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use objc::rc::autoreleasepool;
-use objc::runtime::{Class, Object, BOOL, NO, YES};
-use objc::{class, msg_send, sel, sel_impl};
-use objc_foundation::{INSArray, INSString};
-use objc_foundation::{NSArray, NSString};
-use objc_id::Id;
+use std::panic::RefUnwindSafe;
+use std::panic::UnwindSafe;
+
+use objc2::rc::{autoreleasepool, Id};
+use objc2::runtime::ProtocolObject;
+use objc2::{msg_send_id, ClassType};
+use objc2_app_kit::{NSPasteboard, NSPasteboardTypeFileURL, NSPasteboardTypeString};
+use objc2_foundation::{NSArray, NSString, NSURL};
 
 use crate::common::*;
 
 pub struct OSXClipboardContext {
-    pasteboard: Id<Object>,
+    pasteboard: Id<NSPasteboard>,
 }
 
-// required to bring NSPasteboard into the path of the class-resolver
-#[link(name = "AppKit", kind = "framework")]
-extern "C" {
-    pub static NSPasteboardTypeFileURL: *mut Object;
-    pub static NSPasteboardTypeString: *mut Object;
-}
+unsafe impl Send for OSXClipboardContext {}
+unsafe impl Sync for OSXClipboardContext {}
+impl UnwindSafe for OSXClipboardContext {}
+impl RefUnwindSafe for OSXClipboardContext {}
 
 impl OSXClipboardContext {
     pub fn new() -> Result<OSXClipboardContext> {
-        let cls = Class::get("NSPasteboard").ok_or("Class::get(\"NSPasteboard\")")?;
-        let pasteboard: *mut Object = unsafe { msg_send![cls, generalPasteboard] };
-        if pasteboard.is_null() {
-            return Err("NSPasteboard#generalPasteboard returned null".into());
-        }
-        let pasteboard: Id<Object> = unsafe { Id::from_ptr(pasteboard) };
+        // Use `msg_send_id!` instead of `NSPasteboard::generalPasteboard()`
+        // in the off case that it will return NULL (even though it's
+        // documented not to).
```

Comment:
Does `generalPasteboard` panic if it's null? This sounds like we're just not calling it because it could return null, but so does msg_send_id.

Reply:
`+[NSPasteboard generalPasteboard]` is specified to never return NULL, but since Objective-C is a lot more loose in that regard than Rust, it may still be NULL in reality (e.g. in case of an allocation error). To ensure soundness in the case of a mismatch like this, `objc2-app-kit` will automatically `unwrap` the pointer for you.

If you make `msg_send_id!` return an `Option` instead, it will not unwrap the value automatically, and we get to do so ourselves. This is somewhat documented [here](https://docs.rs/objc2/0.5.1/objc2/macro.msg_send_id.html#specification).

I doubt this error case is common, I mostly did it to keep roughly the same behaviour as the original code.

Reply:
Considering crates shouldn't crash, this is probably the best choice.